Add http-cache intentional bypass/expire/hit/miss to instance close telemetry

erdembayar opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Client.Engineering/issues/1451

Regression? Last working version:

## Description
Add telemetry for http cache hit/miss/bypass data for PMUI.  On my local I can see cache miss (it's due intentional cache bypass) count is quite high. Follow up action would be: Instead of just bypassing cache if we reuse of http cache on disc or [increase cache file life](https://github.com/NuGet/Home/issues/3390) (currently just 30 min), add additional caching, tweak when to bypass cache for better performance.
~~Prep for https://github.com/NuGet/Home/issues/11529 or https://github.com/NuGet/NuGetGallery/issues/8048.~~

**Empty project:**
1. Clean http cache + Open PMUI Install/Browser tabs 
![image](https://user-images.githubusercontent.com/8766776/179880598-8c2440e8-d1c6-48b7-aead-a882af8d76c0.png)

2. After 40 minutes + Open PMUI Install/Browser tabs 

**Another small solution:**
1.  Clean http cache + Open PMUI Install/Browser tabs
![image](https://user-images.githubusercontent.com/8766776/179881137-4f8c6fa7-d6fc-468f-a9a1-f39467cae3f5.png)

![image](https://user-images.githubusercontent.com/8766776/179881234-dac20cb5-1be7-4893-a772-9c34aa982d2f.png)

2. After 40 minutes + Open PMUI Install/Browser tabs

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [x] Test exception: It's bit tedious to add tests here. I manually tested and included screenshots.
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

erdembayar committed:
Add HttpSource usage telemetry

erdembayar committed:
Add intentional PMUI cache pass

erdembayar committed:
Resolve merge conflict

erdembayar committed:
Clean up

erdembayar commented on src/NuGet.Core/NuGet.Protocol/Utility/CachingUtility.cs:
```diff
@@ -50,11 +50,13 @@ public static Stream ReadCacheFile(TimeSpan maxAge, string cacheFile)
+                    return (stream, false);
                 }
+
+                return (null, true);
```
Here `true` means cache expired

erdembayar commented on src/NuGet.Core/NuGet.Protocol/Utility/CachingUtility.cs:
```diff
@@ -50,11 +50,13 @@ public static Stream ReadCacheFile(TimeSpan maxAge, string cacheFile)
                         BufferSize);
 
-                    return stream;
+                    return (stream, false);
```
Here `false` means cache is not expired.

jeffkl replied:
You might as well put a comment in the code here and line 62 with this info

erdembayar commented on src/NuGet.Core/NuGet.Protocol/Utility/CachingUtility.cs:
```diff
@@ -50,11 +50,13 @@ public static Stream ReadCacheFile(TimeSpan maxAge, string cacheFile)
             }
 
-            return null;
+            return (null, false);
```
Cache not found, technically it's same as cache expired.

erdembayar commented on test/TestUtilities/Test.Utility/Protocol/TestHttpSource.cs:
```diff
@@ -62,14 +62,14 @@ public override Task<T> GetAsync<T>(HttpSourceCachedRequest request, Func<HttpSo
             if (DisableCaching)
             {
-                return null;
+                return (null, false);
```
Here explicitly disabling cache, in this case returning `false` for expired property, but this value doesn't matter latter because we have another `bypass` property detect this.

nkolev92 reviewed (changes requested):
Wouldn't cache miss data be more appropriate in a protocol diagnostics event? cc @zivkan wyt?
Please look at the implementation of ProtocolDiagnostics instead.

This PR contains breaking changes that'd need to be reverted as well.

erdembayar committed:
Address PR comment

erdembayar committed:
Reduce Public Api change

erdembayar commented:
> Wouldn't cache miss data be more appropriate in a protocol diagnostics event? cc @zivkan wyt? Please look at the implementation of ProtocolDiagnostics instead.
> 

Moved new cache miss event handler declarations into `ProtocolDiagnostics`.

> This PR contains breaking changes that'd need to be reverted as well.

Reverted existing public api changes.

erdembayar commented on src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSource.cs:
```diff
@@ -111,6 +115,10 @@ public class HttpSource : IDisposable
                     }
+                    else
+                    {
+                        ProtocolDiagnostics.RaiseHttpMissCacheEvent(cacheResult.MaxAge == TimeSpan.Zero, isExpired);
```
Here is `cacheResult.MaxAge == TimeSpan.Zero` means it wants to bypass the cache.

zivkan commented on src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/NuGetHttpCacheUsageCollector.cs:
```diff
@@ -0,0 +1,101 @@
+                }
+            }
+
+            internal void AddProperties(TelemetryEvent telemetryEvent)
+            {
+                telemetryEvent[HttpCache + NuGetHttpCacheUsageCollector.HitCount] = HitCount;
+                telemetryEvent[HttpCache + NuGetHttpCacheUsageCollector.MissCount] = MissCount;
+                telemetryEvent[HttpCache + NuGetHttpCacheUsageCollector.CacheBypassCount] = CacheBypassCount;
+                telemetryEvent[HttpCache + NuGetHttpCacheUsageCollector.ExpiredCacheCount] = ExpiredCacheCount;
+            }
```
> Prep for https://github.com/NuGet/Home/issues/11529, add telemetry for http cache hit/miss data. So, we can answer if additional caching is necessary or tune cache duration time for better performance.

This telemetry won't answer whether or not `If-Modified-Since` will help, since this PR doesn't tell us if the new HTTP response has the same contents as the file in the cache that was not used.

nkolev92 replied:
I agree with @zivkan.

I'm also not sure why this needs a separate event.
http cache hits/misses are on a per project level, so if they were being sent,t hey should be sent through protocol diagnostics.

I do think this information can be valuable, and we might learn something from it, but I don't think it addresses the issue it claims to address.

zivkan replied:
> I'm also not sure why this needs a separate event.
> http cache hits/misses are on a per project level, so if they were being sent,t hey should be sent through protocol diagnostics.

The PR is (now) raising .NET events in the same way that other ProtocolDiagnostics events are (and it already has multiple events). So, FWIW, I'm not 100% sure what you're referring to.

If you mean why is a new .NET event added to the `ProtocolDiagnostics` class, it's consistent with the existing ProtocolDiagnostics architecture. When there's a cache hit `RaiseEvent(ProtocolDiagnosticHttpEvent)` is never called, because it never calls into HttpRetryHandler. `RaiseEvent(ProtocolDiagnosticResourceEvent)` should be called on all SourceRepository resource public APIs, but in order to add cache hit/miss info to those events, HttpSource would have to pass the information all the way up the call stack (and previous comments on this PR already, rightfully, discouraged changs to public APIs.  ProtocolDiagnostics already has 3 different .NET events for this reason - passing data up and down the callstack is difficult and horrible for public APIs, so different events are used.

nkolev92 replied:
I get why there's a new .NET event.

Why shouldn't this be sent with PackageSourceTelemetry?

erdembayar replied:
> 

I'll check if I can add into SourceTelemetry without breaking public apis.

nkolev92 replied:
Personally, I'm not worried about the new public APIs, that seems ok no? 

My comment is mostly whether they should be included in an event about the source usage during that operation. 
I'd wait on a response from @zivkan to make sure we don't want have conflicting feedback before changing too many things.

zivkan replied:
Ah, I get it now. This PR is adding a property to the `vs/nuget/instanceclose` event, but you think it should be added to the `vs/nuget/packagesourcediagnostic` event: https://github.com/NuGet/NuGet.Client/blob/0e3a595fff9f6c0be61132b2771f8401e112b47a/src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs#L220-L255

Yes, that makes sense. We're already collecting a few HTTP related properties there, so it makes sense for us to add it there. Plus, it will allow us to see cache hit/miss differences between PM UI and restore.

However, this "conversation"/thread started because the PR template says that this will help us understand whether `If-Modified-Since` should be implemented, but this PR won't help with that.

erdembayar replied:
@nkolev92 @zivkan 
I did move `http cache logic` to HTTP related properties in `PackageSourceTelemetry`. 
Currently this telemetry is emitted from only 2 places, 1st from [a solution restore job](https://github.com/NuGet/NuGet.Client/blob/f5d7dc371d8a54a9fe0869698e5c4f29dcb7d981/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageFeed.cs#L115) and 2nd from [pmui search](https://github.com/NuGet/NuGet.Client/blob/f5d7dc371d8a54a9fe0869698e5c4f29dcb7d981/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageFeed.cs#L115), both of doesn't happen frequently, so works fine as they happen sequentially.
But it doesn't cover PMUI meta data requests in Install/Update tabs (which consist of hundreds of Http calls), so undercounting one of heavy usage area, fixing this would be quite a lot of refactoring in many places, ROI would be not great, also considering some PMUI design change happening now it's not good timing. I tried to add `PackageSourceTelemetry` in [MultiSourcePackageMetadataProvider](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageMetadataProvider.cs#L256-L267) (which is closest possible place to UI), as [you can see](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageMetadataProvider.cs#L256-L267) each `packageId` meta data requests are handled separately and in it making concurrent http calls to all sources which makes it impossible to use it with `PackageSourceTelemetry`.
Here is example: Let's say Project has following dependencies `PackageA`, `PackageB`, `PackageC` etc..
`PackageA` creates new `PackageSourceTelemetry` then register own [HttpCacheEvent](https://github.com/NuGet/NuGet.Client/blob/dev-eryondon-addcache-hit-miss/src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs#L67).
Same time `PackageB`, `PackageC` the register to same event same time.  Now everytime [HttpCacheEvent](https://github.com/NuGet/NuGet.Client/blob/563d1f2e9f110081b1394e41f9a854e39113f1f4/src/NuGet.Core/NuGet.Protocol/Events/ProtocolDiagnostics.cs#L41) is triggered it send data to all 3 instances of `PackageSourceTelemetry` therefore inflate numbers to all instances listening to it.

So, proposal is "revert back to previous implementation (collect all then send data when close VS, see screenshot in description area)", but I'll group numbers by source and add hash change detection. Does it sound workable solution for you?

nkolev92 replied:
http communication is something that NuGet always does globally. 

I am not sure I follow your concern about double counting. Why is that particular to package source diagnostics and not to any other implementation? 

@zivkan also brought up a very interesting point: 

> However, this "conversation"/thread started because the PR template says that this will help us understand whether If-Modified-Since should be implemented, but this PR won't help with that.

which I agree with. This just captures cache hit/miss but does not really tell us whether something actually changed. 

 > ROI would be not great, also considering some PMUI design change happening now it's not good timing
 
 I think the same logic applies to every attempt to track the PM UI behavior, which what you're trying to do regardless of the implementation. I get that it's slightly different because it'll report it at the end of the process, but I don't think it carries much value when implemented in such a way.

zivkan replied:
Putting aside the issue that this PR will not help the scenarios that the PR description claims

> fixing this would be quite a lot of refactoring in many places, ROI would be not great

On the other hand, HTTP timeouts are NuGet's number 1 unhandled exception in VS, so improving this would benefit not only the telemetry you want to add, but also help us understand what customer experiences are being harmed by HTTP timeouts.

> Now everytime [HttpCacheEvent](https://github.com/NuGet/NuGet.Client/blob/563d1f2e9f110081b1394e41f9a854e39113f1f4/src/NuGet.Core/NuGet.Protocol/Events/ProtocolDiagnostics.cs#L41) is triggered it send data to all 3 instances of PackageSourceTelemetry therefore inflate numbers to all instances listening to it.

~This suggests that you should add `string PackageSource { get; set;}` to to `HttpCacheEvent`, so when the event happens, you know which source it happened in, and you can filter in the event listener. Isn't that what all the other PackageSourceDiagnostics events do?~

No, the code you added already filters to a single source: https://github.com/NuGet/NuGet.Client/pull/4668/files#diff-cd3abb581fb4e39b8057c9bb2da526ee0a16bb3371a049b375e4552ccfb2984aR199-R202

erdembayar replied:
> ~This suggests that you should add `string PackageSource { get; set;}` to to `HttpCacheEvent`, so when the event happens, you know which source it happened in, and you can filter in the event listener. Isn't that what all the other PackageSourceDiagnostics events do?~
> 
> No, the code you added already filters to a single source: https://github.com/NuGet/NuGet.Client/pull/4668/files#diff-cd3abb581fb4e39b8057c9bb2da526ee0a16bb3371a049b375e4552ccfb2984aR199-R202

Main problem is how PMUI acquire meta data here, PMUI get meta data request made 1 by 1 for per package through http starting from PackageItemViewModel level. So now each `packageId + source` is new event, not only they interfere each other, also increase number of `PackageSourceTelemetry` many multifold and spam telemetry service.

erdembayar replied:
@andy @nkolev92 
Yesterday, I accidentally discovered meta data path such as `https://api.nuget.org/v3/registration5-gz-semver2/stackexchange.redis/index.json` are not well tested/supported by nuget.org, by playing with `If-Modified-Since` support for nuget.org in order to detect meta data content change. 

The CDN edge node for [my ISP sometimes doesn't honor](https://github.com/NuGet/NuGetGallery/issues/9175) `If-Modified-Since`, `Etag` header (@joelverhagen can repro too). So, adding content change detection is not reliable at this point with `If-Modified-Since`, `Etag` headers. Fixing that is not high priority for server team right now, if there's any data to prioritize it then they can do it.

At this moment could scope effort on `http-cache bypass/expire/hit/miss` count telemetry for PMUI without content change detection? 

Actually, we can do meta data content change frequency analysis directly on nuget.org, ofc there could be some data correlation issue depending on analysis conclusion.

What we can achieve with this `http-cache intentional bypass/expire/hit/miss` count is:
1. If cache `miss` happening without most of the time without `intentional bypass` then not much we can do, no follow up action.
2. If ton of cache `miss` happening with `intentional bypass` then we should consider making use of existing cache on disc with combination of `If-Modified-Since` or `Etag`. For this we could ask server team to fix it.
3. If a lot of `hit` and `expire` is happening then we should consider [increase cache file life](https://github.com/NuGet/Home/issues/3390) (currently just 30 min) or with combination of `If-Modified-Since` or `Etag` just update the modified datetime of existing cache on disc.  Currently it deletes the existing cache file after 30 min considering it as expired then download exact same file to replace it. For this we could ask server team to fix it.
4. If a lot of hit without expire then probably not much room for improvement. It means most calls are made to different endpoints, caching doesn't work well here.

nkolev92 replied:
> Yesterday, I accidentally discovered meta data path such as https://api.nuget.org/v3/registration5-gz-semver2/stackexchange.redis/index.json are not well tested/supported by nuget.org, by playing with If-Modified-Since support for nuget.org in order to detect meta data content change.

I'm not sure I understand what this means. Registration pages are very much supported and likely tested. 
I guess you mean to say the `If-Modified-Since` header isn't? 

> At this moment could scope effort on http-cache bypass/expire/hit/miss count telemetry for PMUI without content change detection?

That sound good, but it's only 1 of the questions we had. 
The `scope` of the data collected matters too. A few of us voiced feedback about collecting session telemetry only and that it's not helpful. 

> What we can achieve with this http-cache intentional bypass/expire/hit/miss count is:

Thanks for the summary, I think that's reasonable, something that probably should have probably been a telemetry design as @dominoFire suggested.

> If a lot of hit and expire is happening then we should consider https://github.com/NuGet/Home/issues/3390 (currently just 30 min) or with combination of If-Modified-Since or Etag just update the modified datetime of existing cache on disc. Currently it deletes the existing cache file after 30 min considering it as expired then download exact same file to replace it. For this we could ask server team to fix it.

When we make assumptions about what a server can do, we *have* to account for the fact that nuget.org is not the only source. Our solutions need can't rely on nuget.org behavior only.

erdembayar replied:
> > Yesterday, I accidentally discovered meta data path such as https://api.nuget.org/v3/registration5-gz-semver2/stackexchange.redis/index.json are not well tested/supported by nuget.org, by playing with If-Modified-Since support for nuget.org in order to detect meta data content change.
> 
> I'm not sure I understand what this means. Registration pages are very much supported and likely tested. I guess you mean to say the `If-Modified-Since` header isn't?
> 
Yes, I mean `If-Modified-Since` header is not supported well, details are [here](https://github.com/NuGet/NuGetGallery/issues/9175).

> > At this moment could scope effort on http-cache bypass/expire/hit/miss count telemetry for PMUI without content change detection?
> 
> That sound good, but it's only 1 of the questions we had. The `scope` of the data collected matters too. A few of us voiced feedback about collecting session telemetry only and that it's not helpful.
> 

Sure, I'll reevaluate if need to collect new telemetry, technically I already implemented `If-Modified-Since` [here](https://github.com/NuGet/NuGet.Client/pull/4668/files#diff-fe623b43eb533767f04846c3aea5266985147b70a31c2386d7bd8d9275991f63R137-R141). So, this moment just go ahead with implementation without adding new telemetry could be cheaper, but there is some edge case concern, so I'm thinking about writing actual spec for implementing `If-Modified-Since`. Actually, when Joel tested `If-Modified-Since` worked as expected for 20/20. It looks like some CDN edge nodes are misbehaving.

> When we make assumptions about what a server can do, we _have_ to account for the fact that nuget.org is not the only source. Our solutions need can't rely on nuget.org behavior only.

I know many smaller org want to see new packages immediately without caching, so I was not planning to collect data for them. My goal was NuGet.org for initial step extend to Azure and others gradually if it works for NuGet.org. 

I'll step back and reevaluate everything 1 more time.

nkolev92 replied:
> Sure, I'll reevaluate if need to collect new telemetry, technically I already implemented If-Modified-Since [here](https://github.com/NuGet/NuGet.Client/pull/4668/files#diff-fe623b43eb533767f04846c3aea5266985147b70a31c2386d7bd8d9275991f63R137-R141). So, this moment just go ahead with implementation without adding new telemetry could be cheaper, but there is some edge case concern, so I'm thinking about writing actual spec for implementing If-Modified-Since. Actually, when Joel tested If-Modified-Since worked as expected for 20/20. It looks like some CDN edge nodes are misbehaving.

As long as we're not getting false negative we can consider it. 

I'd be ok with implementing the header. I'm likely to consider adding it when working on the vulnerabities in restore.

> I know many smaller org want to see new packages immediately without caching, so I was not planning to collect data for them. My goal was NuGet.org for initial step extend to Azure and others gradually if it works for NuGet.org.

That wasn't really my point. We can make protocol improvements and we can tweak nuget.org, but when we look at our client perf, we can't make any of those assumptions unless we have a special case such as vulnerabilities, because they are *only* reported in nuget.org and we can define support for IF-Modified-Since on that potential new resource as a requiement.

zivkan commented on src/NuGet.Core/NuGet.Protocol/Events/ProtocolDiagnostics.cs:
```diff
@@ -17,6 +17,14 @@ public static class ProtocolDiagnostics
 
         public static event ProtocolDiagnosticsNupkgCopiedEventHandler NupkgCopiedEvent;
 
+        public delegate void HttpCacheHitEventHandler();
+
+        public static event HttpCacheHitEventHandler HttpCacheHitEvent;
+
+        public delegate void HttpCacheMissEventHandler(bool cacheByPass, bool isExpired);
+
+        public static event HttpCacheMissEventHandler HttpCacheMissEvent;
```
This doesn't feel in line with the other ProtocolDiagnostic events. Both hit and miss are cache events, so I think it makes sense to have a `ProtocolDiagnosticHttpCacheEvent`, and then a single new event in this class.

erdembayar committed:
Fix test

erdembayar committed:
Create ProtocolDiagnosticResourceEvent

erdembayar committed:
Add Search meta data

erdembayar committed:
Check redundant listening, inflating number

dominoFire commented:
I would suggest creating a telemetry spec first to understand the needs in a broader picture: https://github.com/NuGet/Client.Engineering/blob/main/docs/telemetry/design-guide.md

erdembayar committed:
Emit 1 telemetry for PMUI

erdembayar committed:
Add operation source

erdembayar committed:
Add IfModifiedSince header

erdembayar committed:
Enroll in experimental flight

erdembayar commented:
> I would suggest creating a telemetry spec first to understand the needs in a broader picture: https://github.com/NuGet/Client.Engineering/blob/main/docs/telemetry/design-guide.md

Sorry about responding to your comments late. By the time I received your comment I was already deep into implementation, in the beginning it was small PR, now it's large. For now, I suspended development, I'll steps back and fill that telemetry spec.

ghost commented:
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 7 days of this comment**. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.

ghost commented:
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 7 days of this comment**. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.
